src/gurt/shm_alloc.c

```diff
@@ -0,0 +1,331 @@
+/**
+ * (C) Copyright 2024 Intel Corporation.
+ * (C) Copyright 2025 Hewlett Packard Enterprise Development LP
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <pthread.h>
+
+#include <gurt/atomic.h>
+#include <gurt/common.h>
+#include <gurt/shm_internal.h>
+#include <gurt/shm_alloc.h>
+#include <gurt/shm_utils.h>
+
+/* the name of shared memory used for mmap which will be found under /dev/shm/ */
+#define daos_shm_name "daos_shm_cache"
+
+/* pid of current process */
+static int          pid;
+
+/* the memory allocator that will be used to handle small memory allocation */
+static __thread int idx_small = -1;
+
+/* the address of shared memory region */
+struct d_shm_hdr   *d_shm_head;
+
+/* the attribute set for mutex located inside shared memory */
+pthread_mutexattr_t d_shm_mutex_attr;
+
+/* the number of times to try calling shm_open() */
+#define RETRY (5)
+
+/**
+ * pid of the process who creates shared memory region. shared memory is NOT unmapped when this
+ * process exits to keep shm always available. shared memory is unmapped when other processes
+ * exit.
+ */
+static int          pid_shm_creator;
+
+static uint64_t     page_size;
+
+static int
+create_shm_region(uint64_t shm_size, uint64_t shm_pool_size)
+{
+	int      i;
+	int      shm_ht_fd;
+	int      shmopen_perm = 0600;
+	void    *shm_addr;
+	char     daos_shm_name_buf[64];
+
+	/* the shared memory only accessible for individual user for now */
+	sprintf(daos_shm_name_buf, "%s_%d", daos_shm_name, getuid());
+	shm_ht_fd = shm_open(daos_shm_name_buf, O_RDWR | O_CREAT | O_EXCL, shmopen_perm);
+	/* failed to create */
+	if (shm_ht_fd == -1) {
+		if (errno == EEXIST)
+			return errno;
+		DS_ERROR(errno, "shm_open() failed to create shared memory");
+		return errno;
+	}
+	/* set the size of shared memory region */
+	if (ftruncate(shm_ht_fd, shm_size) != 0) {
+		DS_ERROR(errno, "ftruncate() failed for shm_ht_fd");
+		goto err;
+	}
+	/* map the shared memory at a fixed address for now. We will remove this limit later. */
```

Comment:
can we add some more details for this in the comment as to why FIXED is needed for the allocator.

Reply:
ok. I will add more details as suggested. Thank you!

Reply:
Fixed mapping address is not needed with updated tlsf memory alllocator.